### PR TITLE
Optimize item.lua

### PR DESCRIPTION
**Pull request by tenplus1**
Replaced slow ItemStack get definitions with faster registered_nodes one's, also cached playername as it's used multiple times and removed 'local item = itemstack:peek_item()' as it is never used.

10,000 loop using ItemStack(node):get_definition() = .2358ms
10,000 loop using registered_nodes[node] = 0.0001ms

### CODE REVIEW

File: builtin/game/item.lua [hunk 1]
```diff
@@ -245,15 +245,15 @@ function core.item_place_node(itemstack, placer, pointed_thing, param2)
 
 	if core.is_protected(place_to, placer:get_player_name()) and
```
Comment (red-001) [inline builtin/game/item.lua]: couldn't you use `playername` here?

File: builtin/game/item.lua [hunk 2]
```diff
@@ -309,7 +309,7 @@ function core.item_place_node(itemstack, placer, pointed_thing, param2)
 
 	-- Run script hook
 	local _, callback
-	for _, callback in ipairs(core.registered_on_placenodes) do
+	for _, callback in pairs(core.registered_on_placenodes) do
```
Comment (Ekdohibs) [inline builtin/game/item.lua]: Why was this changed?

Comment (SmallJoker) [inline builtin/game/item.lua]: `pairs` is slightly faster than `ipairs`. The order doesn't matter here, so this change LGTM.

**Comment by sofar** [comment]
Do you have any data to demonstrate that the array lookup is faster than `:get_definition()`?

File: builtin/game/item.lua [hunk 3]
```diff
@@ -309,7 +309,7 @@ function core.item_place_node(itemstack, placer, pointed_thing, param2)
 
 	-- Run script hook
 	local _, callback
```
Comment (SmallJoker) [inline builtin/game/item.lua]: While we're at optimizing.. Why is this even needed?

File: builtin/game/item.lua (hunk 2 above)
Comment (Ekdohibs) [inline builtin/game/item.lua]: Yes, this was what I was worried about.

**Comment by raymoo** [comment]
The replacement of invocations of get_player_name wouldn't really increase performance, since only at most one of them in the original code would have ended up being called anyway. But it does look cleaner this way.

**Comment by tenplus1** [comment]
I could also remove the ItemStack from line:159 and replace with:
```
local drop = minetest.registered_nodes[nodename].drop
```
... which would be faster for drop lists also.

File: builtin/game/item.lua [hunk 4]
```diff
@@ -156,7 +156,7 @@ function core.yaw_to_dir(yaw)
 end
 
 function core.get_node_drops(nodename, toolname)
-	local drop = ItemStack({name=nodename}):get_definition().drop
+	local drop = core.registered_nodes[nodename].drop
```
Comment (SmallJoker) [inline builtin/game/item.lua]: This would throw an error for any unknown node.
`get_definition` returns the table of athe node `unknown`, thus this needs a crash-safe check.

File: builtin/game/item.lua [hunk 5]
```diff
@@ -449,7 +449,7 @@ function core.handle_node_drops(pos, drops, digger)
 end
 
 function core.node_dig(pos, node, digger)
-	local def = ItemStack({name=node.name}):get_definition()
+	local def = core.registered_nodes[node.name]
```
Comment (SmallJoker) [inline builtin/game/item.lua]: `core.registered_nodes[node.name] or {}`

Comment (tenplus1) [inline builtin/game/item.lua]: Added errorchecking for both.

**Comment by rubenwardy** [comment]
Looks good, I will need to test this however

**Comment by tenplus1** [comment]
@rubenwardy - hows the testing coming along ? can we merge yet ?

**Comment by tenplus1** [comment]
@sfan5 - This was rebased already and |Travis reports everything as ok...  Am I missing something ?

**Comment by sfan5** [comment]
github says this (I am assuming it's accurate):
bildschirmfoto_2017-02-23_12-13-44

**Comment by tenplus1** [comment]
I dont see that, I have a "This branch has no conflicts within the base branch" and a huge greem tick.

**Comment by tenplus1** [comment]
Notice: The last 2 commits were an error and I cannot remove them, help required...

"update lua_api.txt for falling nodes" and "del" should be removed.

**Comment by numberZero** [comment]
@tenplus1 
`git checkout -B patch-1 c99f49f`
`git push --force`

or, if the branch is not checked out,
`git branch --force patch-1 c99f49f`
`git push --force patch-1`

**Comment by tenplus1** [comment]
Thanks @SmallJoker ...
